src/lumigo_tracer/lambda_tracer/lambda_reporter.py

```diff
@@ -168,36 +182,135 @@ def report_json(
+            span_copy.get("info", {}).get("httpInfo", {}).get("response", {}).pop("body", None)
+            return span_copy
+        if span_type == MONGO_SPAN:
+            span_copy.pop("request", None)
+            span_copy.pop("response", None)
+            return span_copy
+        if span_type == REDIS_SPAN:
+            span_copy.pop("requestArgs", None)
+            span_copy.pop("response", None)
+            return span_copy
+        if span_type == SQL_SPAN:
+            span_copy.pop("query", None)
+            span_copy.pop("values", None)
+            span_copy.pop("response", None)
+            return span_copy
+
+    get_logger().warning(f"Got unsupported span type: {span_type}")
+    return span
+
+
 def _create_request_body(
     msgs: List[dict],  # type: ignore[type-arg]
     prune_size_flag: bool,
     max_size: int = MAX_SIZE_FOR_REQUEST,
+    max_error_size: int = MAX_SIZE_FOR_REQUEST_ON_ERROR,
     too_big_spans_threshold: int = TOO_BIG_SPANS_THRESHOLD,
 ) -> str:
+    """
+    This function creates the request body from the given spans.
+    If there is an error we limit the size of the request to max_error_size otherwise to max_size.
+
+    First we try to take all the spans and then we apply the smart span selection.
+
+    The smart span selection has 2 parts:
+    1. We order the spans by FUNCTION_SPAN, ERROR_HTTP_SPAN, ENRICHMENT_SPAN, HTTP_SPAN.
+    2. We take all the spans metadata, We take the full spans. We do this until reach the max_size.
+    """
+    request_size_limit = max_error_size if any(map(is_span_has_error, msgs)) else max_size
 
     if not prune_size_flag or (
         len(msgs) < NUMBER_OF_SPANS_IN_REPORT_OPTIMIZATION
-        and _get_event_base64_size(msgs) < max_size  # noqa
+        and _get_event_base64_size(msgs) < request_size_limit  # noqa
     ):
-        return aws_dump(msgs)[:max_size]
-
-    end_span = msgs[-1]
-    ordered_spans = sorted(msgs[:-1], key=is_span_has_error, reverse=True)
+        return aws_dump(msgs)[:request_size_limit]
 
-    spans_to_send: list = [end_span]  # type: ignore[type-arg]
-    current_size = _get_event_base64_size(end_span)
-    too_big_spans = 0
-    for span in ordered_spans:
+    current_size = 0
+    spans_to_send: List[dict] = []  # type: ignore[type-arg]
+    for index, span in enumerate(msgs):
         span_size = _get_event_base64_size(span)
-        if current_size + span_size < max_size:
-            spans_to_send.append(span)
-            current_size += span_size
-        else:
-            # This is an optimization step. If the spans are too big, don't try to send them.
-            too_big_spans += 1
-            if too_big_spans == too_big_spans_threshold:
-                break
-    return aws_dump(spans_to_send)[:max_size]
+        if current_size + span_size > max_size:
+            break
+
+        spans_to_send.append(span)
+        current_size += span_size
+
+    if len(spans_to_send) < len(msgs):
```

Comment:
note that you're changing the current behavior of the tracer.
To be fair for our users, allow them to skip this update with a configuring env var. Something like "LUMIGO_SPAN_METADATA_UPON_LIMIT" which is True by default

Reply:
I talked with Orr and he don't want to enable the user to do this.
So Im leaving it like this for now.

Reply:
that's surprising..
ok then...